File: leetcode/Convert_Sorted_List_to_Binary_Search_Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    TreeNode sortedListToBST(ListNode head, int n) {
        int m = n / 2;
        ListNode c = head;
        ListNode pre = null;
        int ct = 0;
        while (ct != m) {
            pre = c;
            c = c.next;
            ++ct;
        }
        TreeNode tn = new TreeNode(c.val);
        if (m > 0) {
            pre.next = null;
            tn.left = sortedListToBST(head, m);
        } else {
            tn.left = null;
        }
        if (m < n - 1) {
            tn.right = sortedListToBST(c.next, n - m - 1);
        } else {
            tn.right = null;
        }
        return tn;
    }
    
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) return null;
        int n = 0;
        ListNode c = head;
        while (c != null) {
            ++n;
            c = c.next;
        }
        int m = n / 2;
        c = head;
        ListNode pre = null;
        int ct = 0;
        while (ct != m) {
            pre = c;
            c = c.next;
            ++ct;
        }
        TreeNode tn = new TreeNode(c.val);
        if (m > 0) {
            pre.next = null;
            tn.left = sortedListToBST(head, m);
        } else {
            tn.left = null;
        }
        if (m < n - 1) {
            tn.right = sortedListToBST(c.next, n - m - 1);
        } else {
            tn.right = null;
        }
        return tn;
    }
}

File: leetcode/Convert_Sorted_List_to_Binary_Search_Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int findLength(ListNode *node) {
        int len = 0;
        while (node != NULL) {
            ++len;
            node = node->next;
        }
        return len;
    }
    TreeNode *buildBST(ListNode *node, int len) {
        if (len == 0)
            return NULL;
            
        int mid = 1 + len / 2;
        int idx = 1;
        ListNode *mid_node = node;
        while (idx < mid) {
            mid_node = mid_node->next;
            ++idx;
        }
        
        TreeNode *tree_node = new TreeNode(mid_node->val);
        tree_node->left = buildBST(node, mid - 1);
        tree_node->right = buildBST(mid_node->next, len - mid);
        return tree_node;
    }
public:
    TreeNode *sortedListToBST(ListNode *head) {
        return buildBST(head, findLength(head));
    }
};
